nixos/modules/services/networking/rustical.nix

```diff
@@ -0,0 +1,162 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  inherit (lib)
+    getExe
+    maintainers
+    mkDefault
+    mkEnableOption
+    mkIf
+    mkOption
+    mkPackageOption
+    ;
+
+  inherit (lib.types)
+    attrsOf
+    bool
+    either
+    int
+    nullOr
+    path
+    port
+    str
+    submodule
+    ;
+
+  cfg = config.services.rustical;
+  settingsFormat = pkgs.formats.toml { };
+in
+{
+  options.services.rustical = {
+    enable = mkEnableOption "Rustical CalDAV and CardDAV server";
+    package = mkPackageOption pkgs "rustical" { };
+    settings = mkOption {
+      type = submodule {
+        freeformType = settingsFormat.type;
+        options.http.port = mkOption {
+          type = port;
+          default = 4000;
+          description = ''
+            Which port this service should listen on.
+          '';
+        };
+      };
+      default = { };
+      description = ''
+        Configuration for Rustical, written as TOML
+
+        Refer to the [documentation](https://lennart-k.github.io/rustical/installation/configuration) for options.
+      '';
+    };
+    environmentFile = mkOption {
+      type = nullOr path;
+      default = null;
+      example = "/run/secrets/rustical";
+      description = ''
+        An environment file as defined in {manpage}`systemd.exec(5)`.
+
+        Use this to inject secrets, e.g. database or auth credentials out of band.
+
+        Refer to the [documentation](https://lennart-k.github.io/rustical/installation/configuration/#environment-variables) for options.
+      '';
+    };
+    environment = mkOption {
+      type = attrsOf (either int str);
+      default = { };
+      description = ''
+        Environment variables passed to the service. Any config option name prefixed with RUSTICAL_ takes priority over the one in the configuration file.
+      '';
+      example = {
+        RUSTICAL_DATA_STORE__TOML__PATH = "asd";
+      };
+    };
+    openFirewall = lib.mkOption {
+      type = bool;
+      default = false;
+      description = "Open ports in the firewall for the Rustical.";
+    };
+    user = mkOption {
+      type = str;
+      default = "rustical";
+      description = "User under which the rustical service runs.";
+    };
+    group = mkOption {
+      type = str;
+      default = "rustical";
+      description = "Group under which the rustical service runs.";
+    };
+  };
+
+  config =
+    let
+      dataDir = "/var/lib/rustical";
```

Comment:
non blocking suggestion: it is a good idea to allow configuring datadir via `systemd.tmpfiles.settings`, see for eg.
https://github.com/NixOS/nixpkgs/pull/348588/files#diff-f7430984fbe3b2232ba4d88741af61dd3d92e44de8ae98ffb1bc87dccf842126R107-R112

Reply:
you missed addding your service in nixos/modules/module-list.nix